libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/DataIdHolder.kt

```diff
@@ -0,0 +1,11 @@
+package com.mapbox.navigation.ui.maps.route.line.api
+
+internal class DataIdHolder {
+
+    private val sourceIdToDataIdAndRouteId = mutableMapOf<String, Int>()
+    fun incrementDataId(sourceId: String): Int {
+        val newDataId = (sourceIdToDataIdAndRouteId[sourceId] ?: 0) + 1
+        sourceIdToDataIdAndRouteId[sourceId] = newDataId
+        return newDataId
+    }
```

Comment:
Oh yes, it matters a lot.
TLDR: dataId is an **integer** field that we receive in `onEventDataLoaded` callback if we passed it previously when invoking `featureCollection`. It means that this callback corresponds to this operation. It also means that all previous operations for this source id either already finished or are cancelled. We increment dataId for each operation on this particular sourceId so that we can later keep track of what operations we can consider cancelled.

Reply:
BTW, I'm also sorry that you have to look at this code. :)
It's too complicated for my taste, but I couldn't find a better way. It took me a couple of iterations to a) make the logic consistent; b) simplify the code a bit :) to achieve this.